### PR TITLE
[5.7] Allow overriding Eloquent model array attribute casting

### PR DESCRIPTION
**Pull request by robbytaylor**
The use case is that libraries exist to extend Eloquent for use with other database drivers, such as MongoDB and ElasticSearch. Some of these (in particular MongoDB) have native support for storing arrays without them needing to be encoded as JSON.

Currently, using a MongoDB driver I can't cast an array attribute as such in a model as it results in an error:

```
ErrorException: json_decode() expects parameter 1 to be string, array given
```

I could overload the fromJson method in the MongoDB Eloquent library, but that doesn't seem entirely right. I don't think a method called fromJson should need to check whether the value is already an array.

This would allow Eloquent drivers for database with native array support to override the `fromArray` method to return the value as it is rather than attempting to decode from json.

Admittedly, it isn't actually necessary to use a case at all if the database driver natively supports arrays. But this change makes it easier for the database driver to be changed without changing the casts attribute of the model.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 1]
```diff
@@ -647,6 +648,17 @@ protected function asJson($value)
...
 
+    /**
+     * Alias for fromJson
+     *
+     * @param  string  $value
+     * @return mixed
+     */
+    public function fromArray($value)
+    {
+        return $this->fromJson($value, false);
```
Comment (staudenmeir) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Is `false` necessary? It's already the default value.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 2]
```diff
@@ -481,6 +481,7 @@ protected function castAttribute($key, $value)
             case 'object':
                 return $this->fromJson($value, true);
             case 'array':
+                return $this->fromArray($value);
             case 'json':
                 return $this->fromJson($value);
```
[the remark this answers is not in the document]
Comment (robbytaylor) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: For most situations it's now different from before. It just makes it possible for the fromArray method to be overriden in particular situations (e.g. the database driver supports arrays without storing them as json) without interfering with the json cast.

[the remark this answers is not in the document]
Comment (robclancy) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Honestly there are a lot of places in Laravel that would benefit from the same simple changes so if you have to extend you don't need to copy paste methods.

[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: > For most situations it's no different from before. It just makes it possible for the fromArray method to be overriden in particular situations (e.g. the database driver supports arrays without storing them as json) without interfering with the json cast.

In which case the description "Alias for fromJson." isn't very good.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by robbytaylor** [comment]
My reasoning was that in the HasAttributes trait it is an alias. I've updated the description which hopefully makes the methods intention clearer.

**Comment by crynobone** [comment]
I don't like this as this add a new reserved method in Eloquent.

Would really prefer if we revisit the extendable/instance based casting PR.

**Comment by taylorotwell** [comment]
I agree I don't like adding new reserved methods to Eloquent on a point release, but I also do not totally agree Eloquent should be used to try to shoe-horn other databases into its classes. It was never really designed for that.

IMO other database systems that are not even relational databases should have their own ORMs that fit their particular strengths.

**Pull request closed without merging**